File: game/Baralho.js
// Baralho.js
import { Carta } from './Carta.js';
import { embaralharComRng } from './rng.js';

export class Baralho {
    // numbaralho e a setting randomShuffle vêm do Game via Rodada. `rng` é a
    // fonte de aleatoriedade (ver rng.js): sem seed no Game é Math.random e
    // nada muda; com seed é o PRNG determinístico compartilhado, pra partida
    // reproduzível. Default Math.random pra quem construir Baralho solto.
    constructor(numbaralho, randomShuffle, rng = Math.random) {
        this.numbaralho = numbaralho;
        this.randomShuffle = randomShuffle;
        this.rng = rng;
        this.cartas = []; // O array principal de cartas

        this.montarBaralhos();
    }

    montarBaralhos() {
        let idGeral = 0; // Mantém IDs únicos mesmo em múltiplos baralhos

        for (let b = 0; b < this.numbaralho; b++) {
            // Constrói um baralho de 40 cartas — b+1 porque o baralho é
            // identificado como 1, 2, 3... pro jogador (ver Carta.numeroBaralho)
            let deckAtual = this.construirUmBaralho(idGeral, b + 1);
            idGeral += 40; // Prepara o ID para o próximo baralho não repetir

            if (!this.randomShuffle) {
                // Se NÃO for shuffle global, embaralhamos esse deck sozinho
                this.embaralharArray(deckAtual);

                // Como o comprar() usa pop() (tira do fim do array),
                // os baralhos mais velhos precisam ficar no final.
                // Colocamos o deckAtual NA FRENTE do que já existe em this.cartas
                this.cartas = deckAtual.concat(this.cartas);
            } else {
                // Se for shuffle global, apenas adicionamos ao montante principal
                this.cartas = this.cartas.concat(deckAtual);
            }
        }

        // Se a opção de misturar tudo estiver ativa, embaralha o array inteiro no final
        if (this.randomShuffle) {
            this.embaralharArray(this.cartas);
        }
    }

    // Função separada que retorna um array de 40 cartas.
    // naipes/valores são arrays de [nome, int] (não objeto) de propósito: num
    // objeto, JS reordena chaves que parecem inteiro ('4','5',...,'2','3') pra
    // ordem numérica crescente, então a ordem de construção do monte não seria
    // a escrita aqui — e o motor Python (dict, ordem de inserção) montaria
    // diferente. Como array, os dois montam na mesma ordem, então a mesma seed
    // reproduz a mesma partida nos dois (ver rng.js / training/python/motor/
    // baralho.py).
    construirUmBaralho(idInicial, numeroBaralho) {
        const naipes = [['Ouros', 0], ['Espadas', 1], ['Copas', 2], ['Paus', 3]];
        const valores = [['4', 0], ['5', 1], ['6', 2], ['7', 3], ['Q', 4], ['J', 5], ['K', 6], ['A', 7], ['2', 8], ['3', 9]];

        let deck = [];
        let idCarta = idInicial;

        for (const [nomeNaipe, naipeInt] of naipes) {
            for (const [nomeValor, valorInt] of valores) {
                const novaCarta = new Carta(idCarta, naipeInt, valorInt, nomeNaipe, nomeValor, numeroBaralho);
                deck.push(novaCarta);
                idCarta++;
            }
        }
        return deck;
    }

    embaralharArray(array) {
        embaralharComRng(array, this.rng);
    }

    // Baralho vazio NUNCA é situação normal: o monte é dimensionado exato pra
    // rodada (ver Rodada: numCards = jogadores*round + 1 pra vira,
    // numBaralho = ceil(numCards/40)). Se chegou aqui, o cálculo de
    // numCards/numBaralho quebrou (ou gameOrder/round foi corrompido) — é um
    // bug, não um caso a tratar. Lança pra parar a partida na hora, com
    // contexto, em vez de devolver null e explodir num TypeError adiante.
    comprar() {
        if (this.cartas.length === 0) {
            throw new Error(
                `Baralho vazio ao comprar carta (${this.numbaralho} baralho(s) = ` +
                `${this.numbaralho * 40} cartas montadas). Erro no cálculo de ` +
                `numCards/numBaralho em Rodada.`
            );
        }
        return this.cartas.pop();
    }
}

File: training/python/motor/baralho.py
# baralho.py -- porte direto de game/Baralho.js. A ordem de naipe/valor e a
# lógica de "misturar cada baralho e empilhar o mais novo por cima" (quando
# randomShuffle=False) é copiada tal e qual -- muda o resultado de quem
# compra o quê primeiro, então não é detalhe cosmético.
import random

from .carta import Carta
from .rng import embaralhar_com_rng

# Listas [nome, int] (não dict) pra a ordem de construção do monte ser
# EXATAMENTE esta nos dois motores -- game/Baralho.js usa array pelo mesmo
# motivo (num objeto, JS reordenaria '4','5',...,'2','3' pra ordem numérica).
# Só importa quando há seed (ver rng.py): a mesma seed reproduz a mesma
# partida em JS e Python só se o monte pré-embaralho for idêntico.
NAIPES = [("Ouros", 0), ("Espadas", 1), ("Copas", 2), ("Paus", 3)]
VALORES = [("4", 0), ("5", 1), ("6", 2), ("7", 3), ("Q", 4), ("J", 5), ("K", 6), ("A", 7), ("2", 8), ("3", 9)]


class Baralho:
    def __init__(self, numbaralho, random_shuffle, rng=None):
        self.numbaralho = numbaralho
        self.random_shuffle = random_shuffle
        # rng None => embaralha com random.shuffle (comportamento historico).
        # rng dado (Jogo com seed) => Fisher-Yates explicito bit a bit igual ao
        # de game/Baralho.js, pra a seed reproduzir a mesma sequencia nos dois
        # motores.
        self.rng = rng
        self.cartas = []
        self._montar_baralhos()

    def _embaralhar(self, lista):
        if self.rng is None:
            random.shuffle(lista)
        else:
            embaralhar_com_rng(lista, self.rng)

    def _construir_um_baralho(self, id_inicial, numero_baralho):
        deck = []
        id_carta = id_inicial
        for nome_naipe, naipe_int in NAIPES:
            for nome_valor, valor_int in VALORES:
                deck.append(Carta(id_carta, naipe_int, valor_int, nome_naipe, nome_valor, numero_baralho))
                id_carta += 1
        return deck

    def _montar_baralhos(self):
        id_geral = 0
        for b in range(self.numbaralho):
            deck_atual = self._construir_um_baralho(id_geral, b + 1)
            id_geral += 40
            if not self.random_shuffle:
                self._embaralhar(deck_atual)
                self.cartas = deck_atual + self.cartas
            else:
                self.cartas = self.cartas + deck_atual
        if self.random_shuffle:
            self._embaralhar(self.cartas)

    def comprar(self):
        # Espelha game/Baralho.js: baralho vazio NUNCA é situação normal (o
        # monte é dimensionado exato pra rodada -- ver rodada.py). Se chegou
        # aqui, o cálculo de num_cards/num_baralho quebrou -- é bug, levanta
        # pra parar tudo com contexto em vez de devolver None.
        if not self.cartas:
            raise RuntimeError(
                f"Baralho vazio ao comprar carta ({self.numbaralho} baralho(s) = "
                f"{self.numbaralho * 40} cartas montadas). Erro no cálculo de "
                f"num_cards/num_baralho em Rodada."
            )
        return self.cartas.pop()
